injector/injector.go

```diff
@@ -0,0 +1,312 @@
+	LinearTrackingEvents    map[string][]string
+	NonLinearTrackingEvents map[string][]string
+	CompanionTrackingEvents map[string][]string
+}
+
+type TrackerInjector struct {
+	replacer macros.Replacer
+	events   VASTEvents
+	me       metrics.MetricsEngine
+	provider *macros.MacroProvider
+}
+
+var trimRunes = "\t\r\b\n "
+
+func NewTrackerInjector(replacer macros.Replacer, provider *macros.MacroProvider, events VASTEvents) *TrackerInjector {
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+
+	injectTracker := false
+	injectVideoClicks := false
+	inlineWrapperTagFound := false
+	wrapperTagFound := false
+	impressionTagFound := false
+	errorTagFound := false
+	creativeId := ""
+	isCreative := false
+	companionTagFound := false
+	nonLineaerTagFound := false
+
+	b := strings.NewReader(vastXML)
+	p := xml.NewDecoder(b)
+
+	for {
+		t, err := p.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			switch tt.Name.Local {
+			case "Wrapper":
+				wrapperTagFound = true
+			case "Creative":
+				isCreative = true
+				for _, attr := range tt.Attr {
+					if strings.ToLower(attr.Name.Local) == "adid" {
+						creativeId = attr.Value
+					}
+				}
+			case "Linear":
+				injectVideoClicks = true
+				injectTracker = true
+			case "VideoClicks":
+				injectVideoClicks = false
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				ti.provider.PopulateEventMacros(creativeId, "", "")
+
+				for _, url := range ti.events.VideoClicks {
+					outputXML.WriteString("<ClickTracking><![CDATA[")
+					ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+					outputXML.WriteString("]]></ClickTracking>")
+				}
+
+				continue
+			case "NonLinearAds":
+				injectTracker = true
+			case "TrackingEvents":
+				if isCreative {
+					injectTracker = false
+					encoder.Flush()
+					encoder.EncodeToken(tt)
+					encoder.Flush()
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "tracking", typ)
+						for _, url := range urls {
+							outputXML.WriteString("<Tracking event=\"")
+							outputXML.WriteString(string(typ))
+							outputXML.WriteString("\"><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					continue
+				}
+			}
+
+		case xml.EndElement:
+			switch tt.Name.Local {
+			case "Impression":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				fmt.Println()
```

Comment:
What is the purpose of this print statement?

Reply:
Removed